Parse passive effects directly onto player-specific DBC

gastank opened the pull request:
Currently the various systems we use to parse spell data effects act upon simc object such as `player_t`, `action_t`, `buff_t`. This can lead to several issues including but not limited to: order of operation dependence for values that have both flat add and pct mul applied, implementation dependence when simc objects use different dbc data for the same object value, effect modifying effect requiring individual manual implementation, easy-to-miss unexpected effects in possibly unrelated passives being missed, and others.

The new system directly modifies the player-specific DBC that is currently used for hotfixes, spell data overrides, and trait effect value modification from trait ranks. As these modifications are directly applied to `spell_data_t`, `spelleffect_data_t`, and `spellpower_data_t` they integrate transparently with all existing dbc data access methods and are agnostic to simc object implementations. For effects that don't directly modify the DBC but instead a base attribute of the object, modifications are pre-registered then applied to the object's base members during construction, either directly or via methods such as `action_t::parse_spell_data()`.

**`player_t::parse_passive_effects( const spell_data_t* )`** is the main entry point. Only spells with the `SX_PASSIVE` flag will be parsed, unless `true` is passed as an optional second argument.

Since modifications are directly made onto the DBC (or pre-registered) it's not necessary to wait until object construction to parse, but it is recommended you wait until various named class module spell_t pointer structs are populated by the end of `init_spells()`.

**`player_t::get_passive_value( const spell_data_t&, string )`**
**`player_t::get_passive_value( const spelleffect_data_t&, string )`**
**`player_t::get_passive_value( const spellpower_data_t&, string )`**
These are used to get the final modified value from the modified DBC for the stat specified. Stat strings can be found in `field_type_map` in player.cpp. A 3-value array will be returned, representing `{ original, flat add, pct mult }`. Variables of type `parsed_value_t<T>` can directly be assigned this array.

**`player_t::parse_all_class_passives()`** will automatically parse all class/spec auras, including additional ones added beyond the first. The table of auras can be found in `_class_passives` in `sc_const_data.cpp`.
**`player_t::parse_all_passive_talents()`** will automatically parse all talents active for the current player, including those from the trait hash, manual talent strings, and enable_all_talents.
**`player_t::parse_all_passive_sets()`** will automatically parse all tier set bonuses for the current expansion.
It is recommended you call these three functions for both convenience and to ensure that no errant effects are missed.

**`player_t::deregister_passive_effects( const spell_data_t* )`** will retroactively remove any existing parsing of the spell and prevent all future parsing of the spell.

**`player_t::register_passive_effect_mask( const spell_data_t*, effect_mask_t )`** will allow you to filter out which effects to be parsed. It uses the same `effect_mask_t` syntax used in `parse_effects()`. This will retroactively adjust any existing parsing of the spell. If the spell has not been parsed yet, `parse_passive_effects()` will still need to be called.
```
// skip effect#1 and effect#5 of the spec spell
register_passive_effect_mask( spec_spell, effect_mask_t( true ).disable( 1, 5 ) )
```

**`player_t::register_passive_affect_list( const spell_data_t*, affect_list_t )`** will allow you to modify which spells are affected by the parsed spell. It uses the same `affect_list_t` syntax used in `parse_effects()`. This will retroactively adjust any existing parsing of the spell. If the spell has not been parsed yet, `parse_passive_effects()` will still need to be called. 
```
// also apply family flag 2 to effect#1 of the spec spell
register_passive_affect_list( spec_spell, affect_list_t( 1 ).add_family( 2 ) )
```

**`player_t::register_passive_spell_override( const spell_data_t&, value, field )`**
**`player_t::register_passive_power_override( const spellpower_data_t& value[, field = "cost" ] )`**
**`player_t::register_passive_effect_override( const spelleffect_data_t&, value[, field = "base_value"] )`**
provides a method within the scope of `player_t` that you can use to directly override the values. Unlike the hotfix system, this can be called as a normal `player_t` method allowing you to use any conditional flows as needed. As this directly manipulates the DBC without any processing, it should be called before any other parse methods are used.

**`player_t::disable_class_spell_auto_cloning`** default `false`. To remain agnostic from order of operations, by default all class spells are cloned into the player DBC when spell_data_t pointers are called via `find_spell()` and related methods. By setting this to `true` you can disable this auto cloning. Be warned that this will means that all spell_data_t pointers before parsing will point to the base unmodified DBC, and any modifications from parsing will not apply to values retrieved from those early pointers.

gastank committed:
Parse passive effects directly onto player-specific DBC

Currently the various systems we use to parse spell data effects act upon
simc object such as `player_t`, `action_t``, `buff_t``. This can lead to
several issues including but not limited to: order of operation dependence
for values that have both flat add and pct mul applied, implementation dependence
when simc objects use different dbc data for the same object value, effect
modifying effect requiring individual manual implementation, easy-to-miss
unexpected effects in possibly unrelated passives being missed, and others.

The new system directly modifies the player-specific DBC that is currently
used for hotfixes, spell data overrides, and trait effect value modification
from trait ranks. As these modifications are directly applied to `spell_data_t`,
`spelleffect_data_t`, and `spellpower_data_t` they integrate transparently
with all existing dbc data access methods and are agnostic to simc object
implementations. For effects that don't directly modify the DBC but instead
a base attribute of the object, modifications are pre-registered then applied
to the object's base members during construction, either directly or via
methods such as `action_t::parse_spell_data()`. Furthermore, because parsing
happens once at player-scope, instead of at object-scope repeated for every
created object, there will be some initialization performance improvements.

`player_t::parse_passive_effects( const spell_data_t* )` is the main entry
point. Only spells with the SX_PASSIVE flag will be parsed, unless `true`
is passed as an optional second argument.

Since modifications are directly made onto the DBC (or pre-registered) it's
not necessary to wait until object construction to parse, but it is recommended
you wait until various named class module spell_t pointer structs are populated
by the end of `init_spells()`.

`player_t::get_passive_value( const spell_data_t&, string )`
`player_t::get_passive_value( const spelleffect_data_t&, string )`
`player_t::get_passive_value( const spellpower_data_t&, string )`
These are used to get the final modified value from the modified DBC for
the stat specified. Stat strings can be found in `field_type_map` in player.cpp.
A 3-value array will be returned, representing { orig, flat, pct }. Variables
of type `parsed_value_t<T>` can directly be assigned this array.

`player_t::parse_all_passive_talents()` will automatically parse all talents
active for the current player, including those from the trait hash, manual
talent strings, and enable_all_talents. It is recommended you all this for
both convenience and to ensure that no errant effects are missed.

`player_t::register_passive_effect_mask( const spell_data_t*, effect_mask_t )`
will allow you to filter out which effects to be parsed. It uses the same
`effect_mask_t` syntax used in `parse_effects()`. For example:
`register_passive_effect_mask( spec_spell, effect_mask_t( true ).disable( 1, 5 ) )`
will cause effect#1 and effect#5 of the spec spell to be skipped. Note that
this only registers the effect mask, `parse_passive_effects()` still needs
to be called later.

`player_t::register_passive_affect_list( const spell_data_t* affect_list_t )`
will allow you to modify which spells are affected by the parsed spell. It
uses the same `affect_list_t` syntax used in `parse_effects()`. For example:
`register_passive_affect_list( spec_spell, affect_list_t( 1 ).add_family( 2 ) )`
will cause effect#1 of the spec spell to also apply to spells with family
flag 2. Note that this only registers the effect mask, `parse_passive_effects()`
still needs to be called later.

gastank committed:
fix build errors

gastank committed:
fix typo & warnings

gastank committed:
add option to disable automatic spell cloning

gastank committed:
[Hunter] adapt for parse_passive_effects

gastank committed:
one more hunter fix

nyterage commented on engine/buff/buff.cpp:
```diff
@@ -820,7 +820,10 @@ buff_t* buff_t::set_duration( timespan_t duration )
+      if ( source )
+        base_buff_duration = source->get_passive_value( data(), "duration" );
+      else
+        data().duration();
```
think you forgot the `base_buff_duration` call here

gastank committed:
fix bugs

gastank committed:
[Player] limit enable_all_sets to current expansion tier sets

gastank committed:
add parse_all_passive_sets()

gastank committed:
also clone class label + resource flat mod fix

gastank committed:
[Death Knight] adapt for parse_passive_effects

gastank committed:
use spell data to set action_t::chain_multiplier

gastank committed:
[Demon Hunter] adapt for parse_passive_effects

nyterage commented on engine/class_modules/sc_death_knight.cpp:
```diff
@@ -14252,6 +14209,46 @@ void death_knight_t::init_spells()
   set_icds();
+
+  // Passives that modify effects
+  register_passive_effect_mask( find_spell( 1236355 ), effect_mask_t( true ).disable( 8 ) );
```
This meant to be here? its handled a few lines later in the rider of the apocalypse season 3 set mask.

gastank committed:
remove temporary code used for testing

Saeldur committed:
[Evoker] buff default simplified ilvl

gastank committed:
fix pct reversal, add proc chance, fix chain target/mult

gastank committed:
[Shaman] adapt for parse_passive_effects

gastank committed:
add precision to debug messages

EvanMichaels committed:
[Rogue] Implement Fatebound Coin (Heads) initial damage buff bug

nyterage committed:
[Spell Data] Add forgotten output for subtype 355

github-actions[bot] committed:
Update Generated Files 7b2ca23

nyterage committed:
[Spell Data] Add Descriptions for even more subtypes

nyterage committed:
[Spell Data] Add forgotten `affected_schools()` entry

github-actions[bot] committed:
Update Generated Files 928d87d

gastank committed:
[Shaman] various data parsing bug fixes

1. Static accumulation will correctly use the proc'd ascendance's duration
instead of spell data duration from DRE talent.
2. Tempest AOE further reduced by enhancement spec aura
3. Stormbringer TWW3 2pc applied to pet spells
4. Bonus crit damage from stormcaller properly accounted for along with
elemental fury & primordial fury

gastank committed:
add parse_all_class_passives, more reversals, decouple effect_mask/affect_list

* update _class_passives in sc_const_data.cpp
* allow reversal of modify school, hasted cd, hasted gcd effects
* add parse_all_class_passives to parse spells in _class_passives
* decouple effect_mask registration from parsing; if you register after parsing
it will reverse the newly ignored effects
* decouple apply_list registration from parsing; if you register after parsin
it will reverse the adjusted effects, apply the new affect_list, then reapply
the effects

gastank committed:
Merge branch 'thewarwithin' into parse-passive

gastank committed:
add deregister_passive_effects to remove/block parsing

gastank committed:
[Warrior] adapt for parse_passive_effects

gastank committed:
add E_APPLY_AREA_AURA_PARTY to parsing

gastank committed:
[Warlock] adapt for parse_passive_effects

gastank committed:
fix power & stack parsing

gastank committed:
[Evoker] adapt for parse_passive_effects

gastank committed:
fix power parsing when all powers have auras

gastank committed:
[Priest] adapt for parse_passive_effects

gastank committed:
fix proc chance parsing

gastank committed:
[Rogue] adapt for parse_passive_effects

gastank committed:
adjust base_cost()

gastank committed:
[Monk] adapt to parse_passive_effects

gastank committed:
adjust for not real monk bugs

gastank committed:
[Paladin] adapt to parse_passive_effects

gastank committed:
Merge branch 'thewarwithin' into parse-passive

gastank committed:
remove remaining unused code

gastank committed:
[Mage] adapt for parse_passive_effects

gastank committed:
[Mage] fix defaault value source for lingering embers

gastank committed:
prevent holy armaments from being cast on enemy

gastank committed:
temporary html reporting

gastank committed:
fix gcc build error

gastank committed:
add error message to track down assert

gastank committed:
try again

gastank committed:
don't trigger greater judgement when target is out of combat

gastank committed:
add register_passive_spell_override

gastank committed:
squash warnings

gastank committed:
add comment wall of text

[skip ci]

gastank committed:
add register_passive_power_override

gastank committed:
update comment

[skip ci]
